### PR TITLE
Rapids and GPU accelerated HPO with spot instance support [ +multi-region support ]

### PR DESCRIPTION
**Pull request by miroenev**
*Description of changes:*
Adding a notebook (rapids_sagemaker_hpo.ipynb) which enables RAPIDS GPU accelerated Bring Your Own Container SageMaker HPO workflow. 

Uses the [US Domestic Airline Dataset](https://www.transtats.bts.gov/DatabaseInfo.asp?DB_ID=120&DB_URL=) to set up a flight delay classification ML workflow using either [XGBoost](https://xgboost.readthedocs.io/en/latest/#) or [RandomForest](https://docs.rapids.ai/api/cuml/stable/cuml_blogs.html#tree-and-forest-models) models.

We host multiple sizes of the dataset in a public S3 bucket, but you are welcome to plug in your own dataset.

Can run in multiple code configurations which include:

* `singleCPU`* = [pandas](https://pandas.pydata.org/) + [sklearn](https://scikit-learn.org/stable/)
* `multiCPU`   = [dask](https://dask.org/) + [pandas](https://pandas.pydata.org/) + [sklearn](https://scikit-learn.org/stable/)

* `singleGPU` = [cudf](https://github.com/rapidsai/cudf) + [cuml](https://github.com/rapidsai/cuml)
* `multiGPU`  = [dask](https://dask.org/) + [cudf](https://github.com/rapidsai/cudf) + [cuml](https://github.com/rapidsai/cuml) 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 1]
```diff
@@ -0,0 +1,1819 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 38,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdin",
+     "output_type": "stream",
+     "text": [
+      "confirm test run? [ y / n ] :  y\n"
```
Comment (aaronmarkham) [inline hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1058]: It breaks after this:

```
ClientError: An error occurred (ValidationException) when calling the CreateTrainingJob operation: No S3 objects found under S3 URL "s3://sagemaker-rapids-hpo-us-east-1/1_year" given in input data source. Please ensure that the bucket exists in the selected region (us-west-2), that objects exist under that S3 prefix, and that the role "arn:aws:iam::082907515444:role/team-markham-a2i" has "s3:ListBucket" permissions on bucket "sagemaker-rapids-hpo-us-east-1". Error message from S3: The bucket is in this region: us-east-1. Please use this region to retry the request
```

Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1058]: Thanks Aaron, looking into how to make the S3 bucket (currently in east-1) accessible across multiple regions. Hope to have a follow up shortly.

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 2]
```diff
@@ -0,0 +1,1819 @@
...
+   "metadata": {},
+   "source": [
+    "<img src='img/header.png'>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Hyper Parameter Optimization (HPO) imporves model quality by searching the space of possible 'architecture parameters,' parameters not usually trained during the learning process.  This search can significantly boost model quality relative to default parameters and non-expert tuning; however, the search over architectures can take a very long time on a non-accelerated platform. In this notebook, we containerize a RAPIDS workflow and run Bring-Yor-Own-Container SageMaker HPO to show how we can overcome the computational complexity of model search. \n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:14]: I'll reword this to avoid confusion.

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 3]
```diff
@@ -0,0 +1,1819 @@
...
+    "<span style=\"color:#8735fb; font-size:22pt\"> **Key Choices** </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's go ahead and choose the configuration options for our HPO run.\n",
+    "\n",
+    "Note that we've set the configuration defaults to values that are on the low end, however you are welcome to scale them up. "
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:61]: Modified the text to be more specific and added small and large HPO example configurations.

-----
...
Below are two reference configurations showing a small and a large scale HPO (sized in terms of total experiments/compute). 

The default values in the notebook are set for the small HPO configuration, however you are welcome to scale them up.

> **small HPO**: 1_year, XGBoost, 3 CV folds, singleGPU, max_jobs = 10, max_parallel_jobs = 2

> **large HPO**: 10_year, XGBoost, 10 CV folds, multiGPU, max_jobs = 100, max_parallel_jobs = 10

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb (hunk 1 above)
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1058]: - Replicated the S3 dataset bucket to us-west-2
- Added a region validation check 
- Added documentation to inform users that they need to run in either the us-west-2 or us-east-1 region

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 4]
```diff
@@ -0,0 +1,1819 @@
...
+    "\n",
+    "rapids_base_container = rapids_stable\n",
+    "assert ( rapids_base_container in [ rapids_stable, rapids_nightly ] )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To get things rolling lets make sure we can query our AWS SageMaker execution role and session as well as our account ID and AWS region."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:530]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 5]
```diff
@@ -0,0 +1,1819 @@
...
+    "<span style=\"color:#8735fb; font-size:18pt\"> 2.1.1 - Write Dockerfile </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We write out the Dockerfile in this cell, write it to disk, and in the next cell execute the docker build command. \n",
+    "\n",
+    "First, lets switch our working directory to the location of the Estimator entypoint and library code."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:625]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 6]
```diff
@@ -0,0 +1,1819 @@
...
+   "metadata": {},
+   "source": [
+    "<span style=\"color:#8735fb; font-size:20pt\"> 2.1 - Containerize and Push to ECR </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now lets turn to building our container so that it can integrate with the AWS SageMaker HPO API."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:503]: Done

File: hyperparameter_tuning/rapids_bring_your_own/code/train.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    start_time = time.time()
+
+    # parse inputs and build cluster
+    rapids_sagemaker = RapidsCloudML ( input_args = sys.argv[1:] )
+    
+    try:
+        print( '--- starting workflow --- \n ')
+
+        # [ optional cross-validation] improves robustness/confidence in the best hyper-params
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/code/train.py:40]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 8]
```diff
@@ -0,0 +1,1819 @@
...
+    "From a ML/algorithm perspective, we offer [XGBoost](https://xgboost.readthedocs.io/en/latest/#) and [RandomForest](https://docs.rapids.ai/api/cuml/stable/cuml_blogs.html#tree-and-forest-models) decision tree models which do quite well on this structured dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "algorithm_choice = 'XGBoost'\n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:130]: Yes, done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 9]
```diff
@@ -0,0 +1,1819 @@
...
+    "dataset_directory = '1_year'\n",
+    "\n",
+    "assert( dataset_directory in [ '1_year', '3_year', '10_year'] )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For demo purposes we host these pre-built datasets in a public S3 bucket, you are of course welcome to copy the data into the corresponding region of your compute, or host your own dataset in a bucket of your choosing."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:95]: Performance happens when querying from a S3 bucket (has to be in the same region as the compute per SageMaker requirements). Added clarity around the region options as well as a validation check, see below in the comment where you experienced an error because of a region mis-match.

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 10]
```diff
@@ -0,0 +1,1819 @@
...
+    "* by *scaling across nodes* and running parallel trials on cloud instances.\n",
+    "\n",
+    "By combining these two powers HPO experiments that feel unapproachable and may take multiple days on CPU instances can complete in just hours. For example, we find a <span style=\"color:#8735fb; font-size:14pt\"> **12X** </span> speedup in wall clock time and a <span style=\"color:#8735fb; font-size:14pt\"> **4.5x** </span> reduction in cost when comparing between GPU and CPU [EC2 Spot instances](https://docs.aws.amazon.com/sagemaker/latest/dg/model-managed-spot-training.html) on 100 XGBoost HPO trials using 10 parallel workers on 10 years of the Airline Dataset (~63M flights). For additional details refer to the <a href='#experiments'>end of the notebook</a>."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With all these powerful tools at our disposal, every data scientist should feel empowered to uplevel their model before serving it to the world!"
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:27]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 11]
```diff
@@ -0,0 +1,1819 @@
...
+    "cv_folds = 3\n",
+    "\n",
+    "assert ( cv_folds >= 1 )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<span style=\"color:#8735fb; font-size:18pt\"> [ Code ] </span>"
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:157]: Agreed

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 12]
```diff
@@ -0,0 +1,1819 @@
...
+    "\n",
+    "* `singleCPU`* = [pandas](https://pandas.pydata.org/) + [sklearn](https://scikit-learn.org/stable/)\n",
+    "* `multiCPU`   = [dask](https://dask.org/) + [pandas](https://pandas.pydata.org/) + [sklearn](https://scikit-learn.org/stable/)\n",
+    "\n",
+    "* `singleGPU` = [cudf](https://github.com/rapidsai/cudf) + [cuml](https://github.com/rapidsai/cuml)\n",
+    "* `multiGPU`  = [dask](https://dask.org/) + [cudf](https://github.com/rapidsai/cudf) + [cuml](https://github.com/rapidsai/cuml) \n",
+    "\n",
+    "For some context, cuDF and cuML are part of the RAPIDS library ecosystem. cuDF is a GPU accelerated dataframe library aimed to mirror pandas, while cuML is a GPU accelerated machine-learning library aimed to mirror sklearn.\n",
+    "\n",
+    "Note that the single-CPU option still uses multiple cores in the model training portion of the workflow; however, to unlock parallelism in each stage of the workflow we use [Dask](https://dask.org/). We've integrated all to of these options in the `rapids_cloud_ml.py` file. \n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:174]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 13]
```diff
@@ -0,0 +1,1819 @@
...
+    "<span style=\"color:#8735fb; font-size:18pt\"> [ Code ] </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We enable the option of running different code variations that unlock increasing amounts of parallelism in the compute workflow.\n",
+    "\n",
+    "* `singleCPU`* = [pandas](https://pandas.pydata.org/) + [sklearn](https://scikit-learn.org/stable/)\n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:166]: Yes improved clarity

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 14]
```diff
@@ -0,0 +1,1819 @@
...
+   "source": [
+    "<span style=\"color:#8735fb; font-size:18pt\"> [ Search Ranges and Strategy ] </span>\n",
+    "<a id='strategy-and-param-ranges'></a>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "One of the most important choices when running HPO is to choose the bounds of the hyper-parameter search process. Below we've set the ranges of the hyper-parameters to allow for interesting variation, you are of course welcome to revise these ranges based on domain knowledge especially if you plan to plug in your own dataset. \n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:201]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 15]
```diff
@@ -0,0 +1,1819 @@
...
+   "source": [
+    "<span style=\"color:#8735fb; font-size:18pt\"> [ Compute Platform ] </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Based on the dataset size and compute choice we will try to recommend an instance choice*, you are of course welcome to select alternate configurations. \n",
+    "> e.g., For the 10_year dataset option, we suggest ml.p3.8xlarge GPU instances and ml.m5.24xlarge CPU instances ( we'll need upwards of 200GB CPU RAM during model training)."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:313]: Agreed, done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 16]
```diff
@@ -0,0 +1,1819 @@
...
+    "<span style=\"color:#8735fb; font-size:20pt\"> 2.3 - Test Estimator </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we are ready to test by asking SageMaker to run the BYOContainer logic inside our Estimator. This is a useful step if you've made changes to your custom logic and are interested in making sure everything works before launching a large HPO search. \n",
+    "\n",
+    "> Note: This verification step will use the default hyper-parameter values declared in our custom train code, as SageMaker HPO will not be orchestrating a search for this single run."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1018]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 17]
```diff
@@ -0,0 +1,1819 @@
...
+      "\n",
+      "\u001b[0m\n",
+      "\u001b[34mparsing compute & algorithm choices from job-name...\n",
+      "\n",
+      "  Compute: single-GPU\n",
+      "  Algorithm: XGBoost\n",
+      "  CV_folds: 3\n",
+      "  RAPIDS version: 14\n",
+      "\u001b[0m\n",
+      "\u001b[34mparsing model hyper-parameters from command line arguments...\n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1166]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 18]
```diff
@@ -0,0 +1,1819 @@
...
+   "metadata": {},
+   "source": [
+    "<span style=\"display: block; text-align: center; color:#8735fb; font-size:30pt\"> **3. Run HPO** </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With a working SageMaker Estimator in hand, the hardest part is behind us. In the key choices section we <a href='#strategy-and-param-ranges'>already defined our search strategy and hyper-parameter ranges</a>, so all that remains is to choose a metric to evaluate performance on. For more documentation check out the AWS SageMaker [HyperParameter Tuner documentation](https://sagemaker.readthedocs.io/en/stable/tuner.html)."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1250]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 19]
```diff
@@ -0,0 +1,1819 @@
...
+   ],
+   "source": [
+    "sagemaker.HyperparameterTuningJobAnalytics(tuning_job_name).dataframe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    " For a more in depth look at the HPO process we invite you to check out the <a href='https://github.com/awslabs/amazon-sagemaker-examples/tree/master/hyperparameter_tuning/analyze_results'>HPO_Analyze_TuningJob_Results.ipynb</a> notebook which shows how we can explore interesting things like the <a href='img/results_analysis.png'>impact of each individual hyper-parameter on the performance metric</a>."
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1634]: Done

File: hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb [hunk 20]
```diff
@@ -0,0 +1,1819 @@
...
+   "metadata": {},
+   "source": [
+    "<span style=\"color:#8735fb; font-size:20pt\"> Summary </span>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We've now succesfully built a RAPIDS ML workflow, containerized it (as a SageMaker Estimator), and launched a set of HPO experiments to find the best hyper-paramters for our model.\n",
```
[the remark this answers is not in the document]
Comment (miroenev) [reply hyperparameter_tuning/rapids_bring_your_own/rapids_sagemaker_hpo.ipynb:1723]: Done

**Comment by aaronmarkham** [comment]
> Tested end to end. It requires two points of user input to confirm running. This is neat, but we'll need to address this for future automated testing.

Also, we should look into how to handle the spot instances option during testing. CI can't stall waiting for instances to spin up... the tests will timeout.
